### ISSUE TITLE
Can rendering be limited to 60 fps, but OnUpdate OR Input Events run faster (2x or 3x times per render)?

### ISSUE THREAD

**Issue by ToolmakerSteve** [urho3d/Urho3D#2851]
Testing on Windows 10, via Urho.Net.

I discovered that setting MaxFPS = 60, limits EVERYTHING to 60 fps.
Including the mouse events.

This seems unnecessarily constraining.

We would like to run a higher frame rate for:

* input events (e.g. mouse move).
* Calculations that will affect a future OnUpdate.

I assume we can correct this for calculations, by running them on a timer with a shorter elapsed time. I would make the timer a bit faster than 2x or 3x, but use some mechanism to track whether OnUpdate has happened yet - so I can sync it to OnUpdate if needed.

But mouse events - I currently see no way to get them in at a faster rate than the render cycle.

--------------------------------------------

OR is there a setting somewhere that is LIMITING the rate of mouse events?

I discovered that when MaxFPS = 200, about 1 in 3 OnUpdates do not have ANY mouse event preceding them. (I've been doing tests, where I log events in memory, then dump them afterwards.)

I recorded by moving mouse in circle. With system ticks, distance moved, and whether mouse event occurred in each update cycle. I can see in the data that the updates with no mouse event happen even when mouse is moving fast. So it isn't that "no move occurred". unless the input library is somehow limiting mouse events.

To clarify, I don't need them faster than 180 or 200 fps - I just need to know we can still get them at a high rate, without a glitch, even if something causes a frame to delay. Even if we can't see the events until after the frame delay ends, we would like accurate info about the time and position (for mouse) of user inputs during the delay.

-------------------------------------------

OR can I say MaxFPS = 180, but then in OnUpdate "do nothing" that affects the rendering (on 2 out of 3 frames, exactly), and somehow force exactly 60 fps from the renderer - to avoid wasting time communicating with gpu?

That would give us higher precision on mouse events, and we'd use those two-out-of-three "frames" as "compute only" frames.

Does this make sense? Is it possible?

**Comment by najak3d** [comment]
I'm looking in the code, and I think I see the culprit logic, under Engine.cpp:ApplyFrameLimit().  

I have a proposed change to the code (untested), that might alleviate this issue. 

        for (;;)
        {
            elapsed = frameTimer_.GetUSec(false);
            if (elapsed >= targetMax)
                break;

            **input->Update(); // !!!!!!!!!! ADDED THIS to update Input while we wait to render next frame**

            // Sleep if 1 ms or more off the frame limiting goal
            if (targetMax - elapsed >= 1000LL)
            {
                auto sleepTime = (unsigned)((targetMax - elapsed) / 1000LL);
                Time::Sleep(sleepTime);
            }
        }

**Comment by eugeneko** [comment]
Many SDL events contain timestamps:
![image](https://user-images.githubusercontent.com/8576192/144806924-07c54c22-1cbd-4faa-b3c6-60a46d7c84f1.png)
You can convert it e.g. to ms delay comparing to `Input::Update` and propagate it in the Urho events.

**Comment by 1vanK** [comment]
Pls do not use bug tracker for questions

**Comment by najak3d** [comment]
> Pls do not use bug tracker for questions

The intent here is to raise this as an issue.  The "question here" is "can we supply a fix to this issue with Urho3D"?  The issue is that Urho3D will cap the user-input scan rate effectively at the current FPS of the render loop.   Let's say you are getting 20 FPS (choppy game) -- this isn't too terrible if your User Inputs are still being processed at 100 Hz, instead of 20 Hz.  20 Hz input rate processing is terrible.

It's "bad enough" to have a 20-30 FPS game.... but is playable/tolerable for most.  However, it becomes much more unplayable if in addition to slow FPS, Urho *also* modifies the way the game "responds to User Input" by slowing down the scan rate.

Many games use WSAD keys to apply a "Force" to the player...  If the game hiccups to 10 FPS ... then that 100 msec hiccup will interpret "W-Key-Down" as "being down for 100 msec" even if the player "just pressed it 5 msec ago"... thus modifying the result dramatically.    Stuff like that.

Although for the "happy path" (games that maintain 60 FPS+ on ALL USER DEVICES) works well enough -- the reason this issue hasn't risen-to-the-top yet is that not enough Urho3D-based games are being released and experienced by a large user base, running on devices where Frame rate is slowed, or hiccups.

Our interest in Urho, is in finding the "next winning solution for .NET 3D rendering/games", to be used by professional applications with thousands of real customers.  Our hope is that Urho3D might qualify to be this "winner".

Steve and I, currently have it narrowed down to Urho3D vs. Unity3D.   The big stopper with Unity3D is that they are not open-source, are License/splash-screen-minded, and are not setup to be "used by an App" -- instead, with Unity3D, your only choice is to "create a Unity App" (that must be built/deployed by Unity).   While for Urho, it's much more friendly, and capable of simply being "included into an existing App" for purposes of rendering in 3D.

So we're being extra critical on the deficiencies of Urho3D, hoping to inspire a few upgrades, to help it become "the winner".

Just 1 year ago, UrhoSharp was Microsoft's "recommended solution" for rendering 3D games/apps in Xamarin.   They have since removed this recommendation, as UrhoSharp became unmaintained.

Now we are pressing to help make Urho.NET into the Microsoft solution for MAUI, which *will* be maintained.  In addition, we are also aiming to do the reverse, and make it so that MAUI UI's can be rendered directly into an OpenGL Texture, and thus renderable easily via an Urho3D Sprite, as an in-game UI.

===
In short, our posts here are aimed at being treated as "Issues", and not just "questions".   We are not interested in "how to get user input from Urho", but rather see it as an issue that Urho's Input processing is "deficient" in a way that should be relatively easy to resolve.  

We suggest that Urho provide BOTH options to the end user -- the "accumulated Events" (where it just gives the user ONE event per update frame which is the sum of all small events), but also that Urho provide the user with a way to obtain and handle the individual rapid-events that are happening with SDL, while Urho is busy "sleeping" waiting on "ApplyFrameLimit()" method.   It should be dispatching UI events while it waits to render next frame.  This is an issue.
